### ISSUE TITLE
Timeout on infinite loop sometimes doesn't terminate

### ISSUE THREAD

**Issue by tchajed** [coq/coq#7430]
#### Version

```
$ coqtop -v
The Coq Proof Assistant, version 8.8+alpha (April 2018)
compiled on Apr 21 2018 15:11:27 with OCaml 4.05.0
```

Coq master `dfbf22a098e8`.

#### Operating system

macOS 10.13

#### Description of the problem

Basically, the following code sometimes infinite loops.

The following attempt at a sleep tactic is unreliable. The idea is to repeat a tactic that grows the goal infinitely, then use `timeout` to keep it to 1 second. With an inlined version of `Basics.apply` it becomes fairly unreliable, often taking seemingly infinite time. Using a definition or using `eapply proj1` makes it more reliable but not 100%. Also, adding an additional `Timeout` to the command doesn't solve the problem I believe.

```coq
Tactic Notation "sleep" :=
  try solve [ timeout 1 (repeat eapply (fun (A B : Type) (f : A -> B) (x : A) => f x)) ].

(* this is much more reliable *)
Tactic Notation "sleep'" :=
  try solve [ timeout 1 (repeat eapply proj1) ].

Theorem true1 : True.
Proof.
  sleep.
  idtac "worked".
Abort.
```

You can test this with the following bash one-liner:

```bash
for i in $(seq 1 10); do timeout 2 coqc sleep.v || echo "failed"; done
```

(timeout is from coreutils; use `gtimeout` on macOS if using Homebrew coreutils)

**Comment by LasseBlaauwbroek** [comment]
@ejgallego, since you were involved in #7408, can I ask if there is any chance that this problem can be resolved? It seems that all related issues like #7408, https://github.com/ocaml/ocaml/issues/4127 and https://github.com/ocaml/ocaml/issues/4127 have been resolved. However, I can still reproduce the script above, even on Ocaml 4.11.1 and Coq 8.12. Is it known what the cause of this is?

Thanks

**Comment by ejgallego** [comment]
I can reproduce, but hard to say what is going on here. Actually the timeout seems to happen correctly even in the case that the tactic hangs, so that's a bit bizarre.
